internal/service/advancedclustertpf/resource.go

```diff
@@ -91,19 +93,50 @@ func (r *rs) Read(ctx context.Context, req resource.ReadRequest, resp *resource.
 }
 
 func (r *rs) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
-	var plan TFModel
-	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
-	if resp.Diagnostics.HasError() {
+	var state, plan TFModel
+	diags := &resp.Diagnostics
+	diags.Append(req.Plan.Get(ctx, &plan)...)
+	diags.Append(req.State.Get(ctx, &state)...)
+	if diags.HasError() {
 		return
 	}
-	tfNewModel, shouldReturn := mockedSDK(ctx, &resp.Diagnostics, plan.Timeouts)
+	patchReq := &admin.ClusterDescription20240805{}
+	clusterChanges := conversion.PatchPayloadHasChangesTpf(ctx, diags, &state, &plan, NewAtlasReq, patchReq)
```

Comment:
Good point, some thoughts:
1. Avoid sending updates if they are not needed.
  - This is especially true for resources with multiple endpoints, e.g., if only adv_config has changed we shouldn't send a cluster PATCH (this is done now)
2. Performance due to smaller payloads (minor importance)
3. Debugging help (minor importance) (if there are any errors it can be easier to see what exactly changed)

The downside is the "magic" and the risk of an incomplete solution. But I prefer it to [100s of lines checking for changes](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1735/files#diff-518b3a200e851d92e5dbe0e7882bc66180011a09895ca7b2370d5fd0a3ee5834R154)

Reply:
Yes its definitely great that we get a generic implementation for other resources. I do also like that it relies on conversion functions (to created the update SDK models), so effectively using this function can be an opt-in for each resource.
The point about "This is especially true for resources with multiple endpoints" is the point that has most value IMO, so for resources that consume single update method we might not have this need.